tiledb/api/src/query/subarray.rs

```diff
@@ -33,6 +39,121 @@ pub struct Subarray<'ctx> {
     raw: RawSubarray,
 }
 
+impl<'ctx> Subarray<'ctx> {
+    pub(crate) fn capi(&self) -> *mut ffi::tiledb_subarray_t {
+        *self.raw
+    }
+
+    pub(crate) fn new(context: &'ctx Context, raw: RawSubarray) -> Self {
+        Subarray { context, raw }
+    }
+
+    /// Return all dimension ranges set on the query.
+    pub fn ranges(
+        &self,
+        schema: &'ctx Schema,
+    ) -> TileDBResult<Vec<Vec<Range>>> {
+        let ctx = self.context();
+        let ndims = schema.domain()?.ndim()? as u32;
+
+        let c_ctx = ctx.capi();
+        let c_subarray = self.capi();
+
+        let mut ranges: Vec<Vec<Range>> = Vec::new();
+        for dim_idx in 0..ndims {
+            let mut nranges: u64 = 0;
+            ctx.capi_return(unsafe {
+                ffi::tiledb_subarray_get_range_num(
+                    c_ctx,
+                    c_subarray,
+                    dim_idx,
+                    &mut nranges,
+                )
+            })?;
+
+            let dim = schema.domain()?.dimension(dim_idx)?;
+            let var_sized_dim = dim.is_var_sized()?;
+
+            let mut dim_ranges: Vec<Range> = Vec::new();
+            for rng_idx in 0..nranges {
+                if var_sized_dim {
+                    let mut start_size: u64 = 0;
+                    let mut end_size: u64 = 0;
+                    ctx.capi_return(unsafe {
+                        ffi::tiledb_subarray_get_range_var_size(
+                            c_ctx,
+                            c_subarray,
+                            dim_idx,
+                            rng_idx,
+                            &mut start_size,
+                            &mut end_size,
+                        )
+                    })?;
+
+                    let start =
+                        vec![0u8; start_size as usize].into_boxed_slice();
+                    let end = vec![0u8; end_size as usize].into_boxed_slice();
+
+                    ctx.capi_return(unsafe {
+                        ffi::tiledb_subarray_get_range_var(
+                            c_ctx,
+                            c_subarray,
+                            dim_idx,
+                            rng_idx,
+                            start.as_ptr() as *mut std::ffi::c_void,
+                            end.as_ptr() as *mut std::ffi::c_void,
+                        )
+                    })?;
+
+                    let dtype = dim.datatype()?;
+                    let cvn = dim.cell_val_num()?;
+                    let range =
+                        TypedRange::from_slices(dtype, cvn, &start, &end)?
+                            .range;
+                    dim_ranges.push(range);
+                } else {
+                    let dtype = dim.datatype()?;
+                    let cvn = dim.cell_val_num()?;
+                    let size = match cvn {
```

Comment:
`u32::from(cvn)` is also possible and a bit more concise

Reply:
Alternatively you could do an outer `match` instead of `dim.is_var_sized()`

Reply:
Oh dear. How have I missed that convenience.... I'll got the from route.